From 5fb1dd8b9d64d39f2550b8b18fbe2bc4a4685c0d Mon Sep 17 00:00:00 2001
From: Sudara <sudara@alonetone.com>
Date: Mon, 1 Jul 2024 20:25:05 +0200
Subject: [PATCH 2/2] Do all filter recalculation on the audio thread

---
 Source/PluginProcessor.cpp | 38 +++++++++++++++++++++---
 Source/PluginProcessor.h   | 60 ++++++++++++++++++++------------------
 2 files changed, 66 insertions(+), 32 deletions(-)

diff --git a/Source/PluginProcessor.cpp b/Source/PluginProcessor.cpp
index 1fc6ea4..2c2a41a 100644
--- a/Source/PluginProcessor.cpp
+++ b/Source/PluginProcessor.cpp
@@ -662,6 +662,8 @@ void PolarDesignerAudioProcessor::processBlock(AudioBuffer<float> &buffer, [[may
         updateLatency();
     }
 
+    recomputeFilterCoefficientsIfNeeded();
+
     int numSamples = buffer.getNumSamples();
 
     // create omni and eight signals
@@ -872,11 +874,16 @@ void PolarDesignerAudioProcessor::setStateInformation(const void *data, int size
     repaintDEQ = true;
 }
 
+/* IMPORTANT: parameterChanged *will* be called by both the message thread AND the audio thread.
+ * This varies by DAW, but in general means one should avoid doing anything expensive.
+ * Instead:
+ * * set atomic flags for the UI to pick up on a timer
+ * * set atomic flags for processBlock to check at the start of each block
+*/
 void PolarDesignerAudioProcessor::parameterChanged(const String &parameterID, float newValue) {
     if (parameterID.startsWith("xOverF") && !loadingFile) {
         int idx = parameterID.getTrailingIntValue() - 1;
-        computeFilterCoefficients(idx);
-        initConvolver(idx);
+        recomputeFilterCoefficients[idx] = true;
         repaintDEQ = true;
     } else if (parameterID.startsWith("solo")) {
         soloActive = false;
@@ -890,7 +897,7 @@ void PolarDesignerAudioProcessor::parameterChanged(const String &parameterID, fl
         nBands = static_cast<int> (nBandsPtr->load()) + 1;
         resetXoverFreqs();
         activeBandsChanged = true;
-        computeAllFilterCoefficients();
+        recomputeAllFilterCoefficients = true;
     } else if (parameterID == "proximity") {
         setProxCompCoefficients(proxDistance->load());
     } else if (parameterID == "zeroDelayMode") {
@@ -905,7 +912,7 @@ void PolarDesignerAudioProcessor::parameterChanged(const String &parameterID, fl
                         vtsParams.getParameter("proximity")->convertTo0to1(oldProxDistanceA));
             }
             zeroDelayModeChanged = true;
-            computeAllFilterCoefficients();
+            recomputeAllFilterCoefficients = true;
         } else {
             if (abLayerState == 0 && !abLayerChanged.get()) {
                 oldProxDistanceB = proxDistance->load();
@@ -1033,6 +1040,29 @@ void PolarDesignerAudioProcessor::resetXoverFreqs() {
     }
 }
 
+void PolarDesignerAudioProcessor::recomputeFilterCoefficientsIfNeeded()
+{
+    if (recomputeAllFilterCoefficients.get())
+    {
+        computeAllFilterCoefficients();
+        recomputeAllFilterCoefficients = false;
+
+        // we don't need to go through the individual bands if we're doing it in bulk
+        return;
+    }
+
+    for (int i = 0; i < 4; ++i)
+    {
+        if (recomputeFilterCoefficients[i].get())
+        {
+            computeFilterCoefficients(i);
+            initConvolver(i);
+            recomputeFilterCoefficients[i] = false;
+        }
+    }
+}
+
+
 // compute filter bank filter coeffs and store in firFilterBuffer
 void PolarDesignerAudioProcessor::computeAllFilterCoefficients() {
     for (int i = 0; i < 4; ++i) {
diff --git a/Source/PluginProcessor.h b/Source/PluginProcessor.h
index 927dbf5..d5aba95 100644
--- a/Source/PluginProcessor.h
+++ b/Source/PluginProcessor.h
@@ -2,20 +2,20 @@
  ==============================================================================
  PluginProcessor.h
  Author: Thomas Deppisch & Simon Beck
- 
+
  Copyright (c) 2019 - Austrian Audio GmbH
  www.austrian.audio
- 
+
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
- 
+
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
- 
+
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  ==============================================================================
@@ -105,19 +105,19 @@ public:
     //==============================================================================
     void getStateInformation (MemoryBlock& destData) override;
     void setStateInformation (const void* data, int sizeInBytes) override;
-    
+
     //==============================================================================
     void parameterChanged (const String &parameterID, float newValue) override;
-    
+
     //==============================================================================
     Result loadPreset (const File& presetFile);
     Result savePreset (File destination);
     File getLastDir() {return lastDir;}
     void setLastDir(File newLastDir);
-    
+
     void startTracking(bool trackDisturber);
     void stopTracking(int applyOptimalPattern);
-    
+
     int getNBands() {return nBands;}
     int getSyncChannelIdx() {return static_cast<int>(*syncChannelPtr) + 1;}
     float getXoverSliderRangeStart (int sliderNum);
@@ -126,9 +126,12 @@ public:
     Atomic<bool> activeBandsChanged = true;
     Atomic<bool> zeroDelayModeChanged = true;
     Atomic<bool> ffDfEqChanged = true;
+    std::array<Atomic<bool>,4> recomputeFilterCoefficients;
+    Atomic<bool> recomputeAllFilterCoefficients;
+
     bool getDisturberRecorded() {return disturberRecorded;}
     bool getSignalRecorded() {return signalRecorded;}
-    
+
     void changeAbLayerState();
     bool abLayerState = 1; // 1 = A is active, 0 = B is active
     Identifier saveTree = "save";
@@ -151,13 +154,13 @@ public:
     AudioVisualiserComponent termControlWaveform;
     AudioPlayHead::CurrentPositionInfo info;
     UndoManager undoManager;
-    
+
     // initial xover frequencies for several numbers of bands
     const float INIT_XOVER_FREQS_2B[1] = {1000.0f};
     const float INIT_XOVER_FREQS_3B[2] = {250.0f,3000.0f};
     const float INIT_XOVER_FREQS_4B[3] = {200.0f,1000.0f,5000.0f};
     const float INIT_XOVER_FREQS_5B[4] = {150.0f,600.0f,2600.0f,8000.0f};
-    
+
     // xover ranges for several numbers of bands
     const float XOVER_RANGE_START_2B[1] = {120.0f};
     const float XOVER_RANGE_END_2B[1] = {12000.0f};
@@ -167,31 +170,31 @@ public:
     const float XOVER_RANGE_END_4B[3] = {450.0f, 2500.0f, 12000.0f};
     const float XOVER_RANGE_START_5B[4] = {120.0f, 500.0f, 2200.0f, 7000.0f};
     const float XOVER_RANGE_END_5B[4] = {200.0f, 1100.0f, 4000.0f, 12000.0f};
-    
+
     int getEqState() {return doEq;}
     void setEqState(int idx);
     void setAbLayer(bool state);
     float hzToZeroToOne(int idx, float hz);
     float hzFromZeroToOne(int idx, float val);
     bool zeroDelayModeActive() { return zeroDelayMode->load() > 0.5f; }
-    
+
     void timerCallback() override;
-    
+
 private:
     //==============================================================================
     JUCE_DECLARE_NON_COPYABLE_WITH_LEAK_DETECTOR (PolarDesignerAudioProcessor)
-    
+
     int nBands = 5;
 
     AudioProcessorValueTreeState vtsParams;
     SharedResourcePointer<SharedParams> sharedParams;
 
     static const int N_CH_IN = 2;
-    
+
     // use odd FIR_LEN for even filter order (FIR_LEN = N+1)
     // (lowpass and highpass need even filter order to put a zero at f=0 and f=pi)
     int firLen;
-        
+
     // free field / diffuse field eq
     dsp::Convolution dfEqOmniConv;
     dsp::Convolution dfEqEightConv;
@@ -201,14 +204,14 @@ private:
     dsp::Convolution ffEqEightConv;
     AudioBuffer<float> ffEqOmniBuffer;
     AudioBuffer<float> ffEqEightBuffer;
-    
+
     // proximity compensation filter
     dsp::IIR::Filter<float> proxCompIIR;
-    
+
     // delay (in case of 1 active band)
     Delay delay;
     AudioBuffer<float> delayBuffer;
-    
+
     std::atomic<float>* nBandsPtr;
     std::atomic<float>* syncChannelPtr;
 //    float oldSyncChannelPtr;
@@ -218,10 +221,10 @@ private:
     std::atomic<float>* bandGains[5];
     float oldBandGains[5];
     std::atomic<float>* allowBackwardsPattern;
-    
+
     std::atomic<float>* proxDistance;
     std::atomic<float>* proxOnOff;
-    
+
     std::atomic<float>* zeroDelayMode;
     std::atomic<float>* soloBand[5];
     std::atomic<float>* muteBand[5];
@@ -237,15 +240,15 @@ private:
     bool disturberRecorded;
     bool signalRecorded;
     int nrBlocksRecorded;
-    
+
     float omniSqSumDist[5], eightSqSumDist[5], omniEightSumDist[5],
           omniSqSumSig[5], eightSqSumSig[5], omniEightSumSig[5];
-    
+
     AudioBuffer<float> filterBankBuffer; // holds filtered data, size: N_CH_IN*5
     AudioBuffer<float> firFilterBuffer; // holds filter coefficients, size: 5
     AudioBuffer<float> omniEightBuffer; // holds omni and fig-of-eight signals, size: 2
     dsp::Convolution convolvers[10]; // holds 2*nBands mono convolvers
-    
+
     double currentSampleRate;
     int currentBlockSize;
 
@@ -263,15 +266,16 @@ private:
     void setMaximumSignalPattern();
     void maximizeSigToDistRatio();
     void updateLatency();
-    
+    void recomputeFilterCoefficientsIfNeeded();
+
     // file handling
     File lastDir;
     std::unique_ptr<PropertiesFile> properties;
     const String presetProperties[27] = {"nrActiveBands", "xOverF1", "xOverF2", "xOverF3", "xOverF4", "dirFactor1", "dirFactor2", "dirFactor3", "dirFactor4", "dirFactor5", "gain1", "gain2", "gain3", "gain4", "gain5", "solo1", "solo2", "solo3", "solo4", "solo5", "mute1", "mute2", "mute3", "mute4", "mute5","ffDfEq","proximity"};
-    
+
     static const int FILTER_BANK_NATIVE_SAMPLE_RATE = 48000;
     static const int FILTER_BANK_IR_LENGTH_AT_NATIVE_SAMPLE_RATE = 401;
-    
+
     static const int DF_EQ_LEN = 512;
     static const int FF_EQ_LEN = 512;
     static const int EQ_SAMPLE_RATE = 48000;
-- 
2.38.1

