From 6aa2462b5b7fcd1e7be84fb7549cdf8fe3628baa Mon Sep 17 00:00:00 2001
From: Sudara <sudara@alonetone.com>
Date: Tue, 2 Jul 2024 14:03:58 +0200
Subject: [PATCH 2/2] Change icon color on enablementChanged

---
 resources/customComponents/DirectivityEQ.h | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/resources/customComponents/DirectivityEQ.h b/resources/customComponents/DirectivityEQ.h
index 0182e36..10fceed 100644
--- a/resources/customComponents/DirectivityEQ.h
+++ b/resources/customComponents/DirectivityEQ.h
@@ -184,6 +184,18 @@ class  DirectivityEQ : public Component, private Slider::Listener, private Label
                 bandHandleKnobImageArea.reduce(proportionOfHeight(0.23f), proportionOfHeight(0.23f));
         }
 
+        void enablementChanged() override
+        {
+            if (!isEnabled())
+            {
+               bandHandleKnobImg->replaceColour(Colours::white, Colours::black);
+            }
+            else
+            {
+               bandHandleKnobImg->replaceColour(Colours::black, Colours::white);
+            }
+        }
+
         bool hitTest(int x, int y) override
         {
             Path knobArea;
-- 
2.38.1

