diff --git a/Source/PluginEditor.cpp b/Source/PluginEditor.cpp
index 434e4ea..b810c0f 100644
--- a/Source/PluginEditor.cpp
+++ b/Source/PluginEditor.cpp
@@ -41,7 +41,7 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
 {
 
     nActiveBands = polarDesignerProcessor.getNBands();
-    nBandsBeforeZeroDelayMode = nActiveBands;
+//    nBandsBeforeZeroDelayMode = nActiveBands;
     syncChannelIdx = polarDesignerProcessor.getSyncChannelIdx();
 
     setResizable(true, true);
@@ -1060,27 +1060,27 @@ void PolarDesignerAudioProcessorEditor::buttonClicked (Button* button)
     if ((button == &tmbNrBandsButton[0]) && (button->getToggleState()))
     {
         valueTreeState.getParameter("nrBands")->setValueNotifyingHost(valueTreeState.getParameter("nrBands")->convertTo0to1((0)));
-        nBandsBeforeZeroDelayMode = 1;
+        valueTreeState.getParameter("nrBandsPreZD")->setValue(1);
     }
     if ((button == &tmbNrBandsButton[1]) && (button->getToggleState()))
     {
         valueTreeState.getParameter("nrBands")->setValueNotifyingHost(valueTreeState.getParameter("nrBands")->convertTo0to1((1)));
-        nBandsBeforeZeroDelayMode = 2;
+        valueTreeState.getParameter("nrBandsPreZD")->setValue(2);
     }
     if ((button == &tmbNrBandsButton[2]) && (button->getToggleState()))
     {
         valueTreeState.getParameter("nrBands")->setValueNotifyingHost(valueTreeState.getParameter("nrBands")->convertTo0to1((2)));
-        nBandsBeforeZeroDelayMode = 3;
+        valueTreeState.getParameter("nrBandsPreZD")->setValue(3);
     }
     if ((button == &tmbNrBandsButton[3]) && (button->getToggleState()))
     {
         valueTreeState.getParameter("nrBands")->setValueNotifyingHost(valueTreeState.getParameter("nrBands")->convertTo0to1((3)));
-        nBandsBeforeZeroDelayMode = 4;
+        valueTreeState.getParameter("nrBandsPreZD")->setValue(4);
     }
     if ((button == &tmbNrBandsButton[4]) && (button->getToggleState()))
     {
         valueTreeState.getParameter("nrBands")->setValueNotifyingHost(valueTreeState.getParameter("nrBands")->convertTo0to1((4)));
-        nBandsBeforeZeroDelayMode = 5;
+        valueTreeState.getParameter("nrBandsPreZD")->setValue(5);
     }
 
     if ((button == &tmbSyncChannelButton[0]) && (button->getToggleState()))
@@ -1114,6 +1114,8 @@ void PolarDesignerAudioProcessorEditor::buttonClicked (Button* button)
            (button == &tmbSyncChannelButton[2]) ||
            (button == &tmbSyncChannelButton[3]) ) && (!button->getToggleState())) {
         tmbABButton.setEnabled(true);
+        tmbSyncChannelButton.disableAllButtons();
+        valueTreeState.getParameter("syncChannel")->setValueNotifyingHost(0);
     }
 
     if (button == &tbLoad)
@@ -1237,7 +1239,10 @@ void PolarDesignerAudioProcessorEditor::buttonClicked (Button* button)
     {
         bool isToggled = button->getToggleState();
         button->setToggleState(!isToggled, NotificationType::dontSendNotification);
-        valueTreeState.getParameter("zeroDelayMode")->setValueNotifyingHost(valueTreeState.getParameter("zeroDelayMode")->convertTo0to1((isToggled ? 0.0f : 1.0f))); // !J!
+
+        // !J!
+        valueTreeState.getParameter("zeroDelayMode")->setValueNotifyingHost(valueTreeState.getParameter("zeroDelayMode")->convertTo0to1((isToggled ? 0.0f : 1.0f)));
+
     }
     else if (button == &tmbABButton[0])
     {
@@ -1494,12 +1499,12 @@ void PolarDesignerAudioProcessorEditor::nActiveBandsChanged()
 
 void PolarDesignerAudioProcessorEditor::timerCallback()
 {
+    // filter coefficients are being recomputed
     if (polarDesignerProcessor.recomputeAllFilterCoefficients.get())
     {
-        // timer must wait until recomputeAllFilterCoefficients is false
-        if (polarDesignerProcessor.recomputeAllFilterCoefficients.value)
-            return;
+        return;
     }
+
     if (polarDesignerProcessor.repaintDEQ.get())
     {
         polarDesignerProcessor.repaintDEQ = false;
@@ -1581,27 +1586,32 @@ void PolarDesignerAudioProcessorEditor::timerCallback()
 void PolarDesignerAudioProcessorEditor::zeroDelayModeChange()
 {
     bool zdIsActive = polarDesignerProcessor.zeroDelayModeActive();
+
     tbZeroDelay.setToggleState(zdIsActive, NotificationType::dontSendNotification);
 
     nActiveBands = polarDesignerProcessor.getNBands();
-    unsigned int nActive = nActiveBands;
 
-    if (polarDesignerProcessor.zeroDelayModeActive())
+    unsigned int nNewActive = nActiveBands;
+
+    if (zdIsActive)
     {
-        nActive = 1;
+        nNewActive = 1;
+        valueTreeState.getParameter("nrBands")->setValueNotifyingHost(valueTreeState.getParameter("nrBands")->convertTo0to1((0)));
     }
     else
     {
-        nActive = nBandsBeforeZeroDelayMode;
+        nNewActive = valueTreeState.getParameter("nrBandsPreZD")->getValue();
+//        nNewActive = nBandsBeforeZeroDelayMode;
     }
 
-    valueTreeState.getParameter("nrBands")->setValueNotifyingHost(valueTreeState.getParameter("nrBands")->convertTo0to1((nActive - 1) * 1.0f));
+    // !J!
+    valueTreeState.getParameter("nrBands")->setValueNotifyingHost(valueTreeState.getParameter("nrBands")->convertTo0to1((nNewActive - 1) * 1.0f));
 
     setSideAreaEnabled(!polarDesignerProcessor.zeroDelayModeActive());
-
+/*
     for (unsigned int i = 0; i < 5; i++)
     {
-        if (i < nActive)
+        if (i < nNewActive)
         {
             slDir[i].setEnabled(true);
             slBandGain[i].setEnabled(true);
@@ -1628,7 +1638,7 @@ void PolarDesignerAudioProcessorEditor::zeroDelayModeChange()
             polarPatternVisualizers[i].setVisible(false);
         }
     }
-
+*/
     directivityEqualiser.resetTooltipTexts();
     directivityEqualiser.repaint();
 }
diff --git a/Source/PluginEditor.h b/Source/PluginEditor.h
index 039e28e..c30228e 100644
--- a/Source/PluginEditor.h
+++ b/Source/PluginEditor.h
@@ -84,7 +84,7 @@ public:
 
     void changeListenerCallback(ChangeBroadcaster* source) override;
 
-    unsigned int nBandsBeforeZeroDelayMode;
+//    unsigned int nBandsBeforeZeroDelayMode;
 
 private:
     static const int EDITOR_MIN_WIDTH = 992; // 1194;
diff --git a/Source/PluginProcessor.cpp b/Source/PluginProcessor.cpp
index 2bf5fff..fb64217 100644
--- a/Source/PluginProcessor.cpp
+++ b/Source/PluginProcessor.cpp
@@ -74,6 +74,8 @@ PolarDesignerAudioProcessor::PolarDesignerAudioProcessor() : AudioProcessor(Buse
                                                                     std::make_unique<AudioParameterBool>( ParameterID{"proximityOnOff", PD_PARAMETER_V1}, "ProximityOnOff", false, AudioParameterBoolAttributes().withCategory(AudioProcessorParameter::genericParameter).withStringFromValueFunction([](bool value, [[maybe_unused]] int maximumStringLength) { return value ? "on" : "off"; })),
                                                                     std::make_unique<AudioParameterBool>( ParameterID{"zeroDelayMode", PD_PARAMETER_V1}, "Zero Latency", false, AudioParameterBoolAttributes().withCategory(AudioProcessorParameter::genericParameter).withStringFromValueFunction([](bool value, [[maybe_unused]] int maximumStringLength) { return value ? "on" : "off"; })),
                                                                     std::make_unique<AudioParameterInt>( ParameterID{"syncChannel", PD_PARAMETER_V1}, "Sync to Channel", 0, 4, 0, AudioParameterIntAttributes().withCategory(AudioProcessorParameter::genericParameter).withStringFromValueFunction([](int value, [[maybe_unused]] int maximumStringLength) { return value == 0 ? "none" : String(value); })),
+                                                                     std::make_unique<AudioParameterInt>( ParameterID{"nrBandsPreZD", PD_PARAMETER_V1}, "Nr. of Bands Pre ZeroDelay", 0, 4, 4, AudioParameterIntAttributes().withLabel("").withCategory(AudioProcessorParameter::genericParameter).withStringFromValueFunction([](int value, [[maybe_unused]] int maximumStringLength) { return String(value + 1); }) ),
+
                                                                 }),
                                                             firLen(FILTER_BANK_IR_LENGTH_AT_NATIVE_SAMPLE_RATE),
                                                             dfEqOmniBuffer(1, DF_EQ_LEN),
@@ -441,19 +443,19 @@ void PolarDesignerAudioProcessor::getStateInformation(MemoryBlock &destData) {
     vtsParams.state.setProperty("ffDfEq", var(doEq), nullptr);
     vtsParams.state.setProperty("oldProxDistance", var(oldProxDistance), nullptr);
 
-//    vtsParams.state.setProperty("nrBands", var(nBands * 1.0f), nullptr);
+    vtsParams.state.setProperty("nrBands", var(nBands * 1.0f), nullptr);
 
     if (abLayerState == 1) {
         layerA = vtsParams.copyState();
         doEqA = doEq;
-        nrBandsA = nBands;
+
         if (!exactlyEqual(proxDistance->load(), 0.0f)) {
             oldProxDistanceA = proxDistance->load(); }
     }
     if (abLayerState == 0) {
         layerB = vtsParams.copyState();
         doEqB = doEq;
-        nrBandsB = nBands;
+
         if (!exactlyEqual(proxDistance->load(), 0.0f)) {
             oldProxDistanceB = proxDistance->load();
         }
@@ -461,12 +463,9 @@ void PolarDesignerAudioProcessor::getStateInformation(MemoryBlock &destData) {
 
     layerA.setProperty("ffDfEq", var(doEqA), nullptr);
     layerA.setProperty("oldProxDistance", var(oldProxDistanceA), nullptr);
-    layerA.setProperty("nrBands", var(nrBandsA * 1.0f), nullptr);
 
     layerB.setProperty("ffDfEq", var(doEqB), nullptr);
     layerB.setProperty("oldProxDistance", var(oldProxDistanceB), nullptr);
-    layerB.setProperty("nrBands", var(nrBandsB * 1.0f), nullptr);
-
 
     ValueTree vtsState = vtsParams.copyState();
     ValueTree AState = layerA.createCopy();
@@ -508,14 +507,13 @@ void PolarDesignerAudioProcessor::setStateInformation(const void *data, int size
             oldProxDistance = static_cast<float> (val.getValue());
         }
     }
-    //    nBands = static_cast<unsigned int> (static_cast<int> (nBandsPtr->load()) + 1);
     if (vtsParams.state.hasProperty("nrBands")) {
+        //    nBands = static_cast<unsigned int> (static_cast<int> (nBandsPtr->load()) + 1);
         Value val = vtsParams.state.getPropertyAsValue("nrBands", nullptr);
         if (val.getValue().toString() != "") {
-            nrBandsA = static_cast<int> (val.getValue());
+            nBands = static_cast<int> (val.getValue());
         }
     }
-
     if (layerB.hasProperty("ffDfEq")) {
         Value val = layerB.getPropertyAsValue("ffDfEq", nullptr);
         if (val.getValue().toString() != "") {
@@ -528,12 +526,13 @@ void PolarDesignerAudioProcessor::setStateInformation(const void *data, int size
             oldProxDistanceB = static_cast<float> (val.getValue());
         }
     }
-    if (layerB.hasProperty("nrBands")) {
-        Value val = layerB.getPropertyAsValue("nrBands", nullptr);
-        if (val.getValue().toString() != "") {
-            nrBandsB = static_cast<int> (val.getValue());
-        }
-    }
+
+//    if (layerB.hasProperty("nrBands")) {
+//        Value val = layerB.getPropertyAsValue("nrBands", nullptr);
+//        if (val.getValue().toString() != "") {
+//             = static_cast<int> (val.getValue());
+//        }
+//    }
 
     nBands = static_cast<unsigned int> (static_cast<int> (nBandsPtr->load()) + 1);
 
@@ -565,6 +564,7 @@ void PolarDesignerAudioProcessor::parameterChanged(const String &parameterID, fl
         // trimPosition->load() = newValue; // TODO: update trimSlider position according to automation
     } else if (parameterID == "nrBands") {
         nBands = static_cast<unsigned int> (static_cast<int> (nBandsPtr->load()) + 1);
+
         resetXoverFreqs();
         recomputeAllFilterCoefficients = true;
         activeBandsChanged = true;
@@ -583,26 +583,40 @@ void PolarDesignerAudioProcessor::parameterChanged(const String &parameterID, fl
     } else if (parameterID == "proximity") {
         setProxCompCoefficients(proxDistance->load());
     } else if (parameterID == "zeroDelayMode") {
+
         updateLatency();
 
-        if (exactlyEqual(newValue, 0.0f)) {
-            if (abLayerState == 0) {
-                vtsParams.getParameter("proximity")->setValueNotifyingHost(
-                        vtsParams.getParameter("proximity")->convertTo0to1(oldProxDistanceB));
-            } else {
-                vtsParams.getParameter("proximity")->setValueNotifyingHost(
-                        vtsParams.getParameter("proximity")->convertTo0to1(oldProxDistanceA));
+        // zeroDelayMode is turned off
+        if (exactlyEqual (newValue, 0.0f))
+        {
+            if (abLayerState == 0) // 0 = Layer A, 1 = Layer B
+            {
+                vtsParams.getParameter ("proximity")->setValueNotifyingHost (vtsParams.getParameter ("proximity")->convertTo0to1 (oldProxDistanceB));
             }
+            else
+            {
+                vtsParams.getParameter ("proximity")->setValueNotifyingHost (vtsParams.getParameter ("proximity")->convertTo0to1 (oldProxDistanceA));
+            }
+            vtsParams.getParameter ("nrBands")->setValueNotifyingHost (vtsParams.getParameter ("nrBandsPreZD")->getValue());
+
             recomputeAllFilterCoefficients = true;
             zeroDelayModeChanged = true;
-        } else {
-            if (abLayerState == 0 && !abLayerChanged.get()) {
+        }
+        else
+        // zeroDelayMode is turned on
+        {
+            if ((abLayerState == 0) && !abLayerChanged.get())
+            {
                 oldProxDistanceB = proxDistance->load();
-            } else if (abLayerState == 1 && !abLayerChanged.get()) {
+            }
+            else if ((abLayerState == 1) && !abLayerChanged.get())
+            {
                 oldProxDistanceA = proxDistance->load();
             }
-            vtsParams.getParameter("proximity")->setValueNotifyingHost(
-                    vtsParams.getParameter("proximity")->convertTo0to1(0));
+
+            vtsParams.getParameter ("nrBandsPreZD")->setValueNotifyingHost (nBandsPtr->load() + 1);
+
+            vtsParams.getParameter ("proximity")->setValueNotifyingHost (vtsParams.getParameter ("proximity")->convertTo0to1 (0));
             zeroDelayModeChanged = true;
         }
     } else if (parameterID == "syncChannel" && syncChannelPtr->load() >= 0.5f) {
@@ -658,7 +672,7 @@ void PolarDesignerAudioProcessor::parameterChanged(const String &parameterID, fl
         } else if (parameterID == "proximityOnOff") {
             paramsToSync.proximityOnOff = !exactlyEqual(proxOnOff->load(), 0.0f);
         } else if (parameterID == "zeroDelayMode") {
-            paramsToSync.zeroDelayMode = !exactlyEqual(zeroDelayMode->load(), 0.0f);
+            paramsToSync.zeroDelayMode = exactlyEqual (zeroDelayMode->load(), 0.0f) ? false : true;
         } else if (parameterID.startsWith("gain")) {
             int idx = parameterID.getTrailingIntValue() - 1;
             paramsToSync.gains[idx] = bandGains[idx]->load();
@@ -1374,6 +1388,7 @@ void PolarDesignerAudioProcessor::timerCallback() {
         int ch = (int) syncChannelPtr->load() - 1;
         ParamsToSync &paramsToSync = sharedParams.get().syncParams.getReference(ch);
 
+        // !J! Type cast alias failure?
         if (!exactlyEqual (nBandsPtr->load(), static_cast<float> (paramsToSync.nrActiveBands)))
         {
             vtsParams.getParameter ("nrBands")->setValueNotifyingHost (
@@ -1453,9 +1468,11 @@ void PolarDesignerAudioProcessor::updateLatency() {
 void PolarDesignerAudioProcessor::changeAbLayerState() {
     abLayerChanged = true;
     ffDfEqChanged = true;
+
     if (abLayerState == 0) {
         layerA = vtsParams.copyState();
         doEqA = doEq;
+
         if (!zeroDelayModeActive()) {
             oldProxDistanceA = proxDistance->load();
         }
@@ -1469,14 +1486,16 @@ void PolarDesignerAudioProcessor::changeAbLayerState() {
             oldProxDistance = 0;
         } else {
             oldProxDistance = oldProxDistanceB;
+            nBands = vtsParams.getParameter("nrBandsPreZD")->getValue();
         }
     } else {
         layerB = vtsParams.copyState();
         doEqB = doEq;
 
-        if ( !zeroDelayModeActive() ) {
+        if (!zeroDelayModeActive() ) {
             oldProxDistanceB = proxDistance->load();
         }
+
         readingSharedParams = true;
 
         vtsParams.state = layerA.createCopy();
@@ -1488,11 +1507,11 @@ void PolarDesignerAudioProcessor::changeAbLayerState() {
         }
         else {
             oldProxDistance = oldProxDistanceA;
+            nBands = vtsParams.getParameter("nrBandsPreZD")->getValue();
         }
     }
     vtsParams.state.setProperty("ffDfEq", var(doEq), nullptr);
-    vtsParams.getParameter("proximity")->setValueNotifyingHost(
-            vtsParams.getParameter("proximity")->convertTo0to1(oldProxDistance));
+    vtsParams.getParameter("proximity")->setValueNotifyingHost(vtsParams.getParameter("proximity")->convertTo0to1(oldProxDistance));
     abLayerChanged = false;
 }
 
diff --git a/Source/PluginProcessor.h b/Source/PluginProcessor.h
index 652ee49..a6c2cf5 100644
--- a/Source/PluginProcessor.h
+++ b/Source/PluginProcessor.h
@@ -151,8 +151,8 @@ public:
     int doEq;
     int doEqA;
     int doEqB;
-    int nrBandsA = MAX_NUM_EQS;
-    int nrBandsB = MAX_NUM_EQS;
+//    int oldNrBandsA = MAX_NUM_EQS;
+//    int oldNrBandsB = MAX_NUM_EQS;
     float oldProxDistance;
     float oldProxDistanceA = 0;
     float oldProxDistanceB = 0;
diff --git a/resources/customComponents/MultiTextButton.h b/resources/customComponents/MultiTextButton.h
index 8994eba..fc73531 100644
--- a/resources/customComponents/MultiTextButton.h
+++ b/resources/customComponents/MultiTextButton.h
@@ -62,6 +62,13 @@ public:
         resized();
     }
 
+    void disableAllButtons() {
+        for (int i = 0; i < textButtonArray.size(); i++)
+        {
+            textButtonArray.getUnchecked(i)->setToggleState(false, juce::NotificationType::dontSendNotification);
+        }
+    }
+
     void disableAllButtonsExcept (int btnNr)
     {
         for (int i = 0; i < textButtonArray.size(); i++)
