From 581644b2473cc290beb6949bbcada37b8a9fb1d8 Mon Sep 17 00:00:00 2001
From: Sudara <sudara@alonetone.com>
Date: Tue, 2 Jul 2024 15:20:29 +0200
Subject: [PATCH] PluginEditor changes for melatonin inspector

---
 Source/PluginEditor.cpp | 56 ++++++++++++++++++++---------------------
 Source/PluginEditor.h   | 22 ++++++++--------
 2 files changed, 40 insertions(+), 38 deletions(-)

diff --git a/Source/PluginEditor.cpp b/Source/PluginEditor.cpp
index 20c3820..50e1f64 100644
--- a/Source/PluginEditor.cpp
+++ b/Source/PluginEditor.cpp
@@ -2,20 +2,20 @@
  ==============================================================================
  PluginEditor.cpp
  Author: Thomas Deppisch & Simon Beck
- 
+
  Copyright (c) 2019 - Austrian Audio GmbH
  www.austrian.audio
- 
+
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
- 
+
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
- 
+
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  ==============================================================================
@@ -24,6 +24,8 @@
 #include "PluginProcessor.h"
 #include "PluginEditor.h"
 #include "../resources/customComponents/ImgPaths.h"
+#include "melatonin_inspector/melatonin/helpers/timing.h"
+
 
 //==============================================================================
 PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesignerAudioProcessor& p,
@@ -39,7 +41,6 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
     maxTargetToSpillFlowStarted(false),
     termStage(PolarDesignerAudioProcessorEditor::terminatorStage::DISABLED)
 {
-    openGLContext.attachTo (*getTopLevelComponent());
 
     nActiveBands = currentProcessor.getNBands();
     syncChannelIdx = currentProcessor.getSyncChannelIdx();
@@ -184,10 +185,10 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
     eqColours[2] = mainLaF.polarVisualizerYellow;
     eqColours[3] = mainLaF.polarVisualizerGreen;
     eqColours[4] = mainLaF.polarVisualizerGreenDark;
-    
+
     // directivity eq
     addAndMakeVisible (&directivityEqualiser);
-    
+
     for (int i = 0; i < maxNumberBands; ++i)
     {
         // SOLO button
@@ -196,14 +197,14 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
         tgbSolo[i].addListener (this);
         tgbSolo[i].setButtonText("S");
         tgbSolo[i].setAlwaysOnTop (true);
-        
+
         // MUTE button
         addAndMakeVisible (&tgbMute[i]);
         tgbMuteAtt[i] = std::unique_ptr<ButtonAttachment>(new ButtonAttachment (valueTreeState, "mute" + String(i+1), tgbMute[i]));
         tgbMute[i].addListener (this);
         tgbMute[i].setButtonText("M");
         tgbMute[i].setAlwaysOnTop (true);
-        
+
         // Direction slider
         addAndMakeVisible (&slDir[i]);
         slDirAtt[i] = std::unique_ptr<SliderAttachment>(new SliderAttachment (valueTreeState, "alpha" + String(i+1), slDir[i]));
@@ -211,7 +212,7 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
         slDir[i].addListener (this);
         slDir[i].setTooltipEditable (true);
         slDir[i].setInterceptsMouseClicks(false, true);
-        
+
         // Band Gain slider
         addAndMakeVisible (&slBandGain[i]);
         slBandGainAtt[i] = std::unique_ptr<SliderAttachment>(new SliderAttachment (valueTreeState, "gain" + String(i+1), slBandGain[i]));
@@ -230,10 +231,10 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
 
         // main directivity Equaliser section
         directivityEqualiser.addSliders (eqColours[i], &slDir[i], (i > 0) ? &slCrossoverPosition[i - 1] : nullptr, (i < maxNumberBands - 1) ? &slCrossoverPosition[i] : nullptr, &tgbSolo[i], &tgbMute[i], &slBandGain[i], &polarPatternVisualizers[i]);
-        
+
         if (i == maxNumberBands - 1)
             break; // there is one slCrossoverPosition less than bands
-        
+
         addAndMakeVisible (&slCrossoverPosition[i]);
         slCrossoverAtt[i] = std::unique_ptr<ReverseSlider::SliderAttachment>(new ReverseSlider::SliderAttachment (valueTreeState, "xOverF" + String(i+1), slCrossoverPosition[i]));
         slCrossoverPosition[i].setSliderStyle (Slider::RotaryHorizontalVerticalDrag);
@@ -241,13 +242,13 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
         slCrossoverPosition[i].setVisible(false);
         slCrossoverPosition[i].setInterceptsMouseClicks(false, true);
     }
-    
+
     directivityEqualiser.initValueBox();
 
     addAndMakeVisible (&tbTerminateSpill);
     tbTerminateSpill.setButtonText ("Terminate Spill");
     tbTerminateSpill.addListener (this);
-    
+
     addAndMakeVisible (&tbMaximizeTarget);
     tbMaximizeTarget.setButtonText ("Maximize Target");
     tbMaximizeTarget.addListener (this);
@@ -346,10 +347,10 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
 
     trimSlider.sliderValueSet = [this] { setTrimValue(nActiveBands); };
     trimSlider.sliderReset = [this] { resetTrim(nActiveBands); };
-    
+
     nActiveBandsChanged();
     zeroDelayModeChange();
-    
+
     addAndMakeVisible(&trimSlider);
     trimSlider.addListener(this);
 
@@ -359,7 +360,7 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
     loadSavedPresetsToList();
 
     startTimer (30);
-    
+
     setEqMode();
 }
 
@@ -414,6 +415,8 @@ PolarDesignerAudioProcessorEditor::~PolarDesignerAudioProcessorEditor()
 //==============================================================================
 void PolarDesignerAudioProcessorEditor::paint (Graphics& g)
 {
+    melatonin::ComponentTimer timer { this };
+
     g.fillAll (mainLaF.mainBackground);
 #ifdef AA_DO_DEBUG_PATH
     g.strokePath (debugPath, PathStrokeType (15.0f));
@@ -609,7 +612,7 @@ void PolarDesignerAudioProcessorEditor::resized()
                 //Helpers to calculate position of next polar visualizer
                 polarVisFlexSum += polarVisFlex;
                 prevPolarVisFlex = bandFlexRelToPolarVisComp;
-                //Calculate of the remaining space 
+                //Calculate of the remaining space
                 if (i == nActiveBands - 1)
                     polarVisualizersComponent.items.add(juce::FlexItem().withFlex(1 - polarVisFlexSum));
 
@@ -625,7 +628,7 @@ void PolarDesignerAudioProcessorEditor::resized()
                 //Helpers to calculate position of next gain slider
                 gainSliderFlexSum += gainSliderFlex;
                 prevGainSliderFlex = bandFlex;
-                //Calculate of the remaining space 
+                //Calculate of the remaining space
                 if (i == nActiveBands - 1)
                     gainBandSlidersComponent.items.add(juce::FlexItem().withFlex(1 - gainSliderFlexSum));
             }
@@ -1070,7 +1073,7 @@ void PolarDesignerAudioProcessorEditor::buttonClicked (Button* button)
     {
         valueTreeState.getParameter("nrBands")->setValueNotifyingHost(valueTreeState.getParameter("nrBands")->convertTo0to1((4)));
     }
-    
+
     if ((button == &tmbSyncChannelButton[0]) && (button->getToggleState() > 0.5f))
     {
         tmbSyncChannelButton.disableAllButtonsApartOf(0);
@@ -1150,7 +1153,6 @@ void PolarDesignerAudioProcessorEditor::buttonClicked (Button* button)
         showActiveTerminatorStage(terminatorStage::DISABLED);
         setMainAreaEnabled(true);
         setSideAreaEnabled(true);
-        resized();
     }
     else if (button == &tbMaximizeTarget)
     {
@@ -1159,7 +1161,6 @@ void PolarDesignerAudioProcessorEditor::buttonClicked (Button* button)
         maximizeTarget = true;
         showActiveTerminatorStage(terminatorStage::DISABLED);
         albPlaybackSpill.startAnimation("PLAYBACK SOURCE  ");
-        resized();
     }
     else if (button == &tbMaxTargetToSpill)
     {
@@ -1213,7 +1214,7 @@ void PolarDesignerAudioProcessorEditor::buttonClicked (Button* button)
     else if (button == &tbZeroDelay)
     {
         bool isToggled = button->getToggleState();
-       
+
         button->setToggleState(!isToggled, NotificationType::dontSendNotification);
     }
     else if (button == &tmbABButton[0])
@@ -1456,7 +1457,7 @@ void PolarDesignerAudioProcessorEditor::nActiveBandsChanged()
             polarPatternVisualizers[i].setActive(false);
             polarPatternVisualizers[i].setVisible(false);
             polarPatternVisualizers[i].setToggleState(false, NotificationType::dontSendNotification);
-            
+
             slDir[i].setVisible(false);
             slBandGain[i].setVisible(false);
             tgbSolo[i].setVisible(false);
@@ -1501,8 +1502,8 @@ void PolarDesignerAudioProcessorEditor::timerCallback()
                 setSideAreaEnabled(false);
                 setMainAreaEnabled(false);
                 currentProcessor.startTracking(maximizeTarget ? false : true);
-                albAcquiringTarget.startAnimation("ACQUIRING TARGET   ", 
-                                                  maximizeTarget ? "STOP PLAYBACK WHEN READY TO MAXIMIZE  " 
+                albAcquiringTarget.startAnimation("ACQUIRING TARGET   ",
+                                                  maximizeTarget ? "STOP PLAYBACK WHEN READY TO MAXIMIZE  "
                                                   : "STOP PLAYBACK WHEN READY TO TERMINATE  ");
                 albPlaybackSpill.stopAnimation();
                 resized();
@@ -1855,7 +1856,7 @@ void PolarDesignerAudioProcessorEditor::setSideAreaEnabled(bool set)
     tmbSyncChannelButton[1].setEnabled(set);
     tmbSyncChannelButton[2].setEnabled(set);
     tmbSyncChannelButton[3].setEnabled(set);
-    
+
     tbLoad.setEnabled(set);
     tbSave.setEnabled(set);
     ibEqCtr[0].setEnabled(set);
@@ -2057,4 +2058,3 @@ int PolarDesignerAudioProcessorEditor::getControlParameterIndex (Component& cont
 
     return -1;
 }
-
diff --git a/Source/PluginEditor.h b/Source/PluginEditor.h
index 842ea7d..43b4708 100644
--- a/Source/PluginEditor.h
+++ b/Source/PluginEditor.h
@@ -2,20 +2,20 @@
  ==============================================================================
  PluginEditor.h
  Author: Thomas Deppisch & Simon Beck
- 
+
  Copyright (c) 2019 - Austrian Audio GmbH
  www.austrian.audio
- 
+
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
- 
+
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
- 
+
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  ==============================================================================
@@ -42,6 +42,8 @@
 #include "../resources/customComponents/PresetListBox.h"
 #include "../resources/customComponents/AnimatedLabel.h"
 #include "../resources/customComponents/GainSlider.h"
+#include "melatonin_inspector/melatonin_inspector.h"
+
 
 typedef AudioProcessorValueTreeState::SliderAttachment SliderAttachment;
 typedef AudioProcessorValueTreeState::ButtonAttachment ButtonAttachment;
@@ -106,7 +108,7 @@ private:
     bool presetLoaded = false;
 
     Colour eqColours[5];
- 
+
     TextButton tbLogoAA;
     TitleBarTextLabel titleCompare, titlePreset;
     TextButton titlePresetUndoButton;
@@ -149,7 +151,7 @@ private:
     std::unique_ptr<SliderAttachment> slProximityAtt;
     std::unique_ptr<SliderAttachment> slDirAtt[5];
     std::unique_ptr<ButtonAttachment> tgbSoloAtt[5], tgbMuteAtt[5], tbAllowBackwardsPatternAtt, tbZeroDelayAtt, tgbProxCtrAtt;
-    
+
     DirectivityEQ directivityEqualiser;
     PolarPatternVisualizer polarPatternVisualizers[5];
 
@@ -184,7 +186,7 @@ private:
 #ifdef AA_DO_DEBUG_PATH
     Path debugPath;
 #endif
-    
+
     //==========================================================================
     void nActiveBandsChanged();
     void loadFile();
@@ -201,8 +203,8 @@ private:
     void notifyPresetLabelChange();
 
     void mouseDown(const MouseEvent& event) override;
-    
-    OpenGLContext openGLContext;
-    
+
+    melatonin::Inspector inspector { *this, false };
+
     JUCE_DECLARE_NON_COPYABLE_WITH_LEAK_DETECTOR (PolarDesignerAudioProcessorEditor)
 };
-- 
2.38.1

