From a6f998041d33c2cc2878668d1c31f0807b9139db Mon Sep 17 00:00:00 2001
From: Sudara <sudara@alonetone.com>
Date: Tue, 2 Jul 2024 18:36:32 +0200
Subject: [PATCH 2/2] Temporary fix for app-wide sluggish button drawing

---
 resources/lookAndFeel/MainLookAndFeel.h | 48 ++++++++++++-------------
 1 file changed, 24 insertions(+), 24 deletions(-)

diff --git a/resources/lookAndFeel/MainLookAndFeel.h b/resources/lookAndFeel/MainLookAndFeel.h
index c7c7f13..ef9c0a2 100644
--- a/resources/lookAndFeel/MainLookAndFeel.h
+++ b/resources/lookAndFeel/MainLookAndFeel.h
@@ -74,16 +74,16 @@ public:
         (void)backgroundColour;
         Rectangle<float> buttonArea(0.0f, 0.0f, button.getWidth(), button.getHeight());
 
-        auto austrianAudioLogoImg = juce::Drawable::createFromImageData(BinaryData::austrianAudioLogo_svg, BinaryData::austrianAudioLogo_svgSize);
-        auto loadArrowImg = juce::Drawable::createFromImageData(BinaryData::loadArrow_svg, BinaryData::loadArrow_svgSize);
-        auto freeFieldImg = juce::Drawable::createFromImageData(BinaryData::freeField_svg, BinaryData::freeField_svgSize);
-        auto diffuseFieldImg = juce::Drawable::createFromImageData(BinaryData::diffuseField_svg, BinaryData::diffuseField_svgSize);
-        auto eqFieldCheckSign = juce::Drawable::createFromImageData(BinaryData::eqFieldCheckSign_svg, BinaryData::eqFieldCheckSign_svgSize);
-        auto terminateSpillIconImg = juce::Drawable::createFromImageData(BinaryData::terminateSpillIcon_svg, BinaryData::terminateSpillIcon_svgSize);
-        auto maximizeTargetIconImg = juce::Drawable::createFromImageData(BinaryData::maximizeTargetIcon_svg, BinaryData::maximizeTargetIcon_svgSize);
-        auto maxTargetToSpillIconImg = juce::Drawable::createFromImageData(BinaryData::maxTargetToSpillIcon_svg, BinaryData::maximizeTargetIcon_svgSize);
-        auto closePresetListIconImg = juce::Drawable::createFromImageData(BinaryData::closePresetListIcon_svg, BinaryData::closePresetListIcon_svgSize);
-        auto undoPresetIconImg = juce::Drawable::createFromImageData(BinaryData::undoPresetIcon_svg, BinaryData::undoPresetIcon_svgSize);
+        static auto austrianAudioLogoImg = juce::Drawable::createFromImageData(BinaryData::austrianAudioLogo_svg, BinaryData::austrianAudioLogo_svgSize);
+        static auto loadArrowImg = juce::Drawable::createFromImageData(BinaryData::loadArrow_svg, BinaryData::loadArrow_svgSize);
+        static auto freeFieldImg = juce::Drawable::createFromImageData(BinaryData::freeField_svg, BinaryData::freeField_svgSize);
+        static auto diffuseFieldImg = juce::Drawable::createFromImageData(BinaryData::diffuseField_svg, BinaryData::diffuseField_svgSize);
+        static auto eqFieldCheckSign = juce::Drawable::createFromImageData(BinaryData::eqFieldCheckSign_svg, BinaryData::eqFieldCheckSign_svgSize);
+        static auto terminateSpillIconImg = juce::Drawable::createFromImageData(BinaryData::terminateSpillIcon_svg, BinaryData::terminateSpillIcon_svgSize);
+        static auto maximizeTargetIconImg = juce::Drawable::createFromImageData(BinaryData::maximizeTargetIcon_svg, BinaryData::maximizeTargetIcon_svgSize);
+        static auto maxTargetToSpillIconImg = juce::Drawable::createFromImageData(BinaryData::maxTargetToSpillIcon_svg, BinaryData::maximizeTargetIcon_svgSize);
+        static auto closePresetListIconImg = juce::Drawable::createFromImageData(BinaryData::closePresetListIcon_svg, BinaryData::closePresetListIcon_svgSize);
+        static auto undoPresetIconImg = juce::Drawable::createFromImageData(BinaryData::undoPresetIcon_svg, BinaryData::undoPresetIcon_svgSize);
 
         int cornerDirButtons = button.getTopLevelComponent()->proportionOfHeight(0.004f);
         int reduceYDirButtons = button.getTopLevelComponent()->proportionOfHeight(0.005f);
@@ -97,7 +97,7 @@ public:
         {
             g.setColour(button.isEnabled() ? textButtonActiveFrameColor : textButtonHoverBackgroundColor);
             g.drawRect(buttonArea, 1);
-            
+
             if (isMouseOverButton)
             {
                 g.setColour(textButtonHoverBackgroundColor);
@@ -230,7 +230,7 @@ public:
         {
             g.setColour(textButtonFrameColor);
             g.drawRect(buttonArea, 1);
-            
+
             if (isMouseOverButton)
             {
                 g.setColour(textButtonHoverBackgroundColor);
@@ -241,7 +241,7 @@ public:
                 g.setColour(textButtonPressedBackgroundColor);
                 g.fillRect(buttonArea.reduced(1.0f, 1.0f));
             }
-            
+
             if (!button.isEnabled())
             {
                 bool resultMainImg = diffuseFieldImg->replaceColour(Colours::white, mainTextDisabledColor);
@@ -259,7 +259,7 @@ public:
             diffuseFieldImageArea.removeFromTop(button.proportionOfHeight(0.11f));
             diffuseFieldImageArea.removeFromBottom(button.proportionOfHeight(0.33f));
             diffuseFieldImg->drawWithin(g, diffuseFieldImageArea, juce::RectanglePlacement::centred, 1.f);
-            
+
             if (button.getToggleState() == true)
             {
                 auto eqFieldCheckSignArea = juce::Rectangle<float>(button.getWidth()*0.77f,
@@ -309,7 +309,7 @@ public:
                     return;
             }
             maximizeTargetIconImg->drawWithin(g, iconArea, juce::RectanglePlacement::centred, 1.f);
-             
+
             if (isMouseOverButton)
             {
                 g.setColour(textButtonHoverBackgroundColor);
@@ -388,14 +388,14 @@ public:
                 ellipseArea.setHeight(jmin(reducedButtonArea.getWidth(), reducedButtonArea.getHeight()));
                 ellipseArea.setCentre(buttonArea.getCentre());
                 path.addEllipse(ellipseArea);
-                
+
                 auto color = textButtonPressedBackgroundColor;
                 if (button.getToggleState())
                 {
                     //Draw check sign when terminator stage completed
-                    Line<float> line1(Point(ellipseArea.getX() + ellipseArea.proportionOfWidth(0.24f), 
-                        ellipseArea.getY() + ellipseArea.proportionOfHeight(0.47f)), 
-                        Point(ellipseArea.getX() + ellipseArea.proportionOfWidth(0.41f), 
+                    Line<float> line1(Point(ellipseArea.getX() + ellipseArea.proportionOfWidth(0.24f),
+                        ellipseArea.getY() + ellipseArea.proportionOfHeight(0.47f)),
+                        Point(ellipseArea.getX() + ellipseArea.proportionOfWidth(0.41f),
                             ellipseArea.getY() + ellipseArea.proportionOfHeight(0.67f)));
 
                     Line<float> line2(line1.getEnd(),
@@ -434,7 +434,7 @@ public:
         {
             g.setColour(mainTextColor);
             g.drawRect(buttonArea, 1);
-            
+
             if (isMouseOverButton)
             {
                 g.setColour(textButtonHoverBackgroundColor);
@@ -654,7 +654,7 @@ public:
             path.lineTo(point4);
             path.lineTo(point5);
             path.closeSubPath();
-            
+
             g.setColour(textButtonFrameColor);
             g.fillPath(path);
         }
@@ -1015,7 +1015,7 @@ public:
         g.setFont(font);
 
         if (button.getButtonText() == "S" || button.getButtonText() == "M")
-        { 
+        {
             toggleButtonBounds.reduce(4, 4);
             if (button.getToggleState())
             {
@@ -1130,7 +1130,7 @@ public:
         {
             p.addEllipse(w*0.23f, h / 2 - newDiameter / 2, newDiameter, newDiameter);
         }
-        
+
         g.setColour(component.isEnabled() ? mainTextColor : mainTextDisabledColor);
         g.fillPath(p);
     }
@@ -1169,7 +1169,7 @@ public:
             g.setColour(labelBackgroundColor);
 
             int deltaX = 0;
-            
+
             if (SystemStats::getOperatingSystemName() == "iOS")
                 deltaX = static_cast<int> (buttonArea.proportionOfWidth (0.24f));
 
-- 
2.38.1

