diff --git a/.gitmodules b/.gitmodules
index bbf7cc5..060c565 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -9,7 +9,3 @@
 [submodule "modules/melatonin_inspector"]
 	path = modules/melatonin_inspector
 	url = https://github.com/sudara/melatonin_inspector.git
-[submodule "modules/melatonin_perfetto"]
-	path = modules/melatonin_perfetto
-	url = https://github.com/sudara/melatonin_perfetto.git
-	branch = main
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 73233bc..bbc1d59 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -62,9 +62,6 @@ add_subdirectory(JUCE)
 # Only linked when #define AA_INCLUDE_MELATONIN
 #add_subdirectory (modules/melatonin_inspector)
 
-# Sudara's Perfetto implementation for JUCE
-add_subdirectory(modules/melatonin_perfetto)
-
 # See `docs/CMake API.md` in the JUCE repo for all config options
 juce_add_plugin("${PROJECT_NAME}"
     # Icons for the standalone app
@@ -170,8 +167,6 @@ target_compile_definitions(SharedCode
     BUILD_COMMIT_HASH="${GIT_COMMIT_HASH}"
     BUILD_TAG="${GIT_TAG}"
 
-	PERFETTO=1
-
     # JucePlugin_Name is for some reason doesn't use the nicer PRODUCT_NAME
     PRODUCT_NAME_WITHOUT_VERSION="PolarDesigner"
 )
@@ -184,8 +179,7 @@ target_compile_definitions(SharedCode
 target_link_libraries(SharedCode
     INTERFACE
     Assets
-#    melatonin_inspector	
-    melatonin_perfetto
+#    melatonin_inspector
     juce_audio_basics
 	juce_audio_devices
 	juce_audio_formats
diff --git a/Source/PluginEditor.cpp b/Source/PluginEditor.cpp
index 584d036..fe6cd86 100644
--- a/Source/PluginEditor.cpp
+++ b/Source/PluginEditor.cpp
@@ -342,12 +342,15 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
     lbFactoryPresets.setHeaderText("Factory Presets");
     lbFactoryPresets.addChangeListener(this);
 
-    nEditorBandsChanged();
-    zeroLatencyModeChanged();
+    updateUIForNumberActiveEQBands();
+    updateUIForZeroLatencyMode();
 
     trimSlider.sliderValueSet = [this] { setTrimValue((int) nEditorBands); };
     trimSlider.sliderReset = [this] { resetTrim((int) nEditorBands); };
 
+    updateUIForNumberActiveEQBands();
+    updateUIForZeroLatencyMode();
+
     addAndMakeVisible(&trimSlider);
     trimSlider.addListener(this);
 
@@ -356,12 +359,9 @@ PolarDesignerAudioProcessorEditor::PolarDesignerAudioProcessorEditor (PolarDesig
 
     loadSavedPresetsToList();
 
-    nEditorBandsChanged();
-    zeroLatencyModeChanged();
-
     startTimer (30);
 
-    setEqMode();
+    updateUIForEQMode();
 }
 
 // Handle the trimSlider increment/decrement calls
@@ -1167,8 +1167,8 @@ void PolarDesignerAudioProcessorEditor::buttonClicked (Button* button)
         uiMaximizeTargetWindowIsVisible = false;
         uiMaxTargetToSpillFlowStarted = false;
         showActiveTerminatorStage(terminatorStage::DISABLED);
-        setMainAreaEnabled(true);
-        enableLatencyUI (true);
+        setMainUIEnabled (true);
+        setLatencyUIEnabled (true);
     }
     else if (button == &tbMaximizeTarget)
     {
@@ -1213,8 +1213,8 @@ void PolarDesignerAudioProcessorEditor::buttonClicked (Button* button)
         uiMaxToSpillWindowIsVisible = false;
         uiMaxTargetToSpillFlowStarted = false;
         uiTargetAquisitionWindowIsVisible = false;
-        enableLatencyUI (true);
-        setMainAreaEnabled(true);
+        setLatencyUIEnabled (true);
+        setMainUIEnabled (true);
         polarDesignerProcessor.stopTracking(2);
         termStage = terminatorStage::DISABLED;
         showActiveTerminatorStage(termStage);
@@ -1232,8 +1232,8 @@ void PolarDesignerAudioProcessorEditor::buttonClicked (Button* button)
     {
         // Toggle the UI
         button->setToggleState(!button->getToggleState(), NotificationType::dontSendNotification);
+        // !J! Disabled while debugging sync issues
         bool isToggled = button->getToggleState();
-
         valueTreeState.getParameter("zeroLatencyMode")->setValueNotifyingHost(
             valueTreeState.getParameter("zeroLatencyMode")->convertTo0to1(isToggled)); // !J!
 
@@ -1407,12 +1407,12 @@ void PolarDesignerAudioProcessorEditor::loadFile()
         Result result = polarDesignerProcessor.loadPreset (presetFile);
         if (!result.wasOk()) {
             errorMessage = result.getErrorMessage();
-            setMainAreaEnabled(false);
-            enableLatencyUI (false);
+            setMainUIEnabled (false);
+            setLatencyUIEnabled (false);
         }
         else
         {
-            setEqMode();
+            updateUIForEQMode();
         }
         loadingFile = false;
     }
@@ -1430,8 +1430,8 @@ void PolarDesignerAudioProcessorEditor::saveFile()
         Result result = polarDesignerProcessor.savePreset (presetFile);
         if (!result.wasOk()) {
             errorMessage = result.getErrorMessage();
-            setMainAreaEnabled(false);
-            enableLatencyUI (false);
+            setMainUIEnabled (false);
+            setLatencyUIEnabled (false);
         }
         else
         {
@@ -1464,7 +1464,7 @@ void PolarDesignerAudioProcessorEditor::loadSavedPresetsToList()
     }
 }
 
-void PolarDesignerAudioProcessorEditor::nEditorBandsChanged()
+void PolarDesignerAudioProcessorEditor::updateUIForNumberActiveEQBands()
 {
     nEditorBands = polarDesignerProcessor.getNProcessorBands();
 
@@ -1508,8 +1508,6 @@ void PolarDesignerAudioProcessorEditor::nEditorBandsChanged()
 
 void PolarDesignerAudioProcessorEditor::timerCallback()
 {
-    TRACE_COMPONENT();
-
     // timer must wait until recomputeAllFilterCoefficients is false
     if (polarDesignerProcessor.recomputeAllFilterCoefficients.get())
     {
@@ -1523,17 +1521,17 @@ void PolarDesignerAudioProcessorEditor::timerCallback()
     if (polarDesignerProcessor.activeBandsChanged.get())
     {
         polarDesignerProcessor.activeBandsChanged = false;
-        nEditorBandsChanged();
+        updateUIForNumberActiveEQBands();
     }
     if (polarDesignerProcessor.zeroLatencyModeChanged.get())
     {
         polarDesignerProcessor.zeroLatencyModeChanged = false;
-        zeroLatencyModeChanged();
+        updateUIForZeroLatencyMode();
     }
     if (polarDesignerProcessor.ffDfEqChanged.get())
     {
         polarDesignerProcessor.ffDfEqChanged = false;
-        setEqMode();
+        updateUIForEQMode();
     }
 
     if (uiTerminatorAnimationWindowIsVisible)
@@ -1543,8 +1541,8 @@ void PolarDesignerAudioProcessorEditor::timerCallback()
             if (!uiTargetAquisitionWindowIsVisible)
             {
                 uiTargetAquisitionWindowIsVisible = true;
-                enableLatencyUI (false);
-                setMainAreaEnabled(false);
+                setLatencyUIEnabled (false);
+                setMainUIEnabled (false);
                 polarDesignerProcessor.startTracking(uiMaximizeTargetWindowIsVisible ? false : true);
                 albAcquiringTarget.startAnimation("ACQUIRING TARGET   ",
                     uiMaximizeTargetWindowIsVisible ? "STOP PLAYBACK WHEN READY TO MAXIMIZE  "
@@ -1558,8 +1556,8 @@ void PolarDesignerAudioProcessorEditor::timerCallback()
             if (uiTargetAquisitionWindowIsVisible)
             {
                 uiTargetAquisitionWindowIsVisible = false;
-                enableLatencyUI (true);
-                setMainAreaEnabled(true);
+                setLatencyUIEnabled (true);
+                setMainUIEnabled (true);
                 polarDesignerProcessor.stopTracking(1);
                 albAcquiringTarget.stopAnimation();
                 albPlaybackSpill.stopAnimation();
@@ -1583,8 +1581,8 @@ void PolarDesignerAudioProcessorEditor::timerCallback()
         if (uiTargetAquisitionWindowIsVisible)
         {
             uiTargetAquisitionWindowIsVisible = false;
-            enableLatencyUI (true);
-            setMainAreaEnabled(true);
+            setLatencyUIEnabled (true);
+            setMainUIEnabled (true);
             polarDesignerProcessor.stopTracking(0);
             albAcquiringTarget.stopAnimation();
             albPlaybackSpill.stopAnimation();
@@ -1593,18 +1591,15 @@ void PolarDesignerAudioProcessorEditor::timerCallback()
     }
 }   
 
-void PolarDesignerAudioProcessorEditor::zeroLatencyModeChanged()
+void PolarDesignerAudioProcessorEditor::updateUIForZeroLatencyMode()
 {
     bool zdIsAlreadyActive = polarDesignerProcessor.zeroLatencyModeActive();
 
-    TRACE_COMPONENT();
-
     tbZeroLatency.setToggleState(zdIsAlreadyActive, NotificationType::dontSendNotification);
     nEditorBands = polarDesignerProcessor.getNProcessorBands();
 
-    /* !J! TODO: DELETE
+    unsigned int mNBandsToSet; // = nEditorBands;
 
-    unsigned int mNBandsToSet = nEditorBands; // = nEditorBands;
     if (polarDesignerProcessor.zeroLatencyModeActive())
     {
         mNBandsToSet = 1;
@@ -1613,12 +1608,12 @@ void PolarDesignerAudioProcessorEditor::zeroLatencyModeChanged()
     else {
         mNBandsToSet = nEditorBands;
     }
-*/
-    enableLatencyUI (!polarDesignerProcessor.zeroLatencyModeActive());
 
-    for (unsigned int i = 0; i < 5; i++)
+    setLatencyUIEnabled (!polarDesignerProcessor.zeroLatencyModeActive());
+
+    for (unsigned int i = 0; i < MAX_EDITOR_BANDS; i++)
     {
-        if (i < nEditorBands)
+        if (i < mNBandsToSet)
         {
             slDir[i].setEnabled(true);
             slBandGain[i].setEnabled(true);
@@ -1876,76 +1871,76 @@ void PolarDesignerAudioProcessorEditor::changeListenerCallback(ChangeBroadcaster
     }
 }
 
-void PolarDesignerAudioProcessorEditor::setMainAreaEnabled(bool enable)
+void PolarDesignerAudioProcessorEditor::setMainUIEnabled (bool enabled)
 {
-    directivityEqualiser.setActive(enable);
-    directivityEqualiser.setEnabled(enable);
+    directivityEqualiser.setActive(enabled);
+    directivityEqualiser.setEnabled(enabled);
 
     for (unsigned int i = 0; i < nEditorBands; i++)
     {
-        directivityEqualiser.getDirPathComponent(static_cast<int> (i)).setEnabled(enable);
-        slDir[i].setEnabled(enable);
-        slBandGain[i].setEnabled(enable);
-        tgbSolo[i].setEnabled(enable);
-        tgbMute[i].setEnabled(enable);
-        polarPatternVisualizers[i].setEnabled(enable);
-        polarPatternVisualizers[i].setActive(enable);
+        directivityEqualiser.getDirPathComponent(static_cast<int> (i)).setEnabled(enabled);
+        slDir[i].setEnabled(enabled);
+        slBandGain[i].setEnabled(enabled);
+        tgbSolo[i].setEnabled(enabled);
+        tgbMute[i].setEnabled(enabled);
+        polarPatternVisualizers[i].setEnabled(enabled);
+        polarPatternVisualizers[i].setActive(enabled);
         if (polarPatternVisualizers[i].getToggleState())
             polarPatternVisualizers[i].setToggleState(false, NotificationType::dontSendNotification);
     }
-    tbZeroLatency.setEnabled(enable);
-    titlePreset.setEnabled(enable);
+    tbZeroLatency.setEnabled(enabled);
+    titlePreset.setEnabled(enabled);
     repaint();
 }
 
-void PolarDesignerAudioProcessorEditor::enableLatencyUI (bool set)
+void PolarDesignerAudioProcessorEditor::setLatencyUIEnabled (bool enabled)
 {
-    tmbNrBandsButton[0].setEnabled (set);
-    tmbNrBandsButton[1].setEnabled (set);
-    tmbNrBandsButton[2].setEnabled (set);
-    tmbNrBandsButton[3].setEnabled (set);
-    tmbNrBandsButton[4].setEnabled (set);
-    tmbSyncChannelButton[0].setEnabled (set);
-    tmbSyncChannelButton[1].setEnabled (set);
-    tmbSyncChannelButton[2].setEnabled (set);
-    tmbSyncChannelButton[3].setEnabled (set);
-
-    tbLoad.setEnabled (set);
-    tbSave.setEnabled (set);
-    ibEqCtr[0].setEnabled (set);
-    ibEqCtr[1].setEnabled (set);
-    tbAllowBackwardsPattern.setEnabled (set);
-    tbTerminateSpill.setEnabled (set);
-    tbMaximizeTarget.setEnabled (set);
-    tbMaxTargetToSpill.setEnabled (set);
-    tgbProxCtr.setEnabled (set);
+    tmbNrBandsButton[0].setEnabled (enabled);
+    tmbNrBandsButton[1].setEnabled (enabled);
+    tmbNrBandsButton[2].setEnabled (enabled);
+    tmbNrBandsButton[3].setEnabled (enabled);
+    tmbNrBandsButton[4].setEnabled (enabled);
+    tmbSyncChannelButton[0].setEnabled (enabled);
+    tmbSyncChannelButton[1].setEnabled (enabled);
+    tmbSyncChannelButton[2].setEnabled (enabled);
+    tmbSyncChannelButton[3].setEnabled (enabled);
+
+    tbLoad.setEnabled (enabled);
+    tbSave.setEnabled (enabled);
+    ibEqCtr[0].setEnabled (enabled);
+    ibEqCtr[1].setEnabled (enabled);
+    tbAllowBackwardsPattern.setEnabled (enabled);
+    tbTerminateSpill.setEnabled (enabled);
+    tbMaximizeTarget.setEnabled (enabled);
+    tbMaxTargetToSpill.setEnabled (enabled);
+    tgbProxCtr.setEnabled (enabled);
     if (tgbProxCtr.getToggleState())
     {
-        slProximity.setEnabled (set);
+        slProximity.setEnabled (enabled);
     }
-    grpEq.setEnabled (set);
-    grpPreset.setEnabled (set);
-    grpProxComp.setEnabled (set);
-    grpBands.setEnabled (set);
-    grpSync.setEnabled (set);
-    grpPresetList.setEnabled (set);
+    grpEq.setEnabled (enabled);
+    grpPreset.setEnabled (enabled);
+    grpProxComp.setEnabled (enabled);
+    grpBands.setEnabled (enabled);
+    grpSync.setEnabled (enabled);
+    grpPresetList.setEnabled (enabled);
 
     if (polarDesignerProcessor.zeroLatencyModeActive()) {
-        grpTerminatorControl.setEnabled (set);
+        grpTerminatorControl.setEnabled (enabled);
     }
     else {
         grpTerminatorControl.setEnabled (true);
     }
 
-    tmbABButton.setEnabled(set);
-    tmbABButton.setVisible(set);
-    titlePreset.setEnabled(set);
-    titleCompare.setEnabled(set);
+    tmbABButton.setEnabled(enabled);
+    tmbABButton.setVisible(enabled);
+    titlePreset.setEnabled(enabled);
+    titleCompare.setEnabled(enabled);
 
     repaint();
 }
 
-void PolarDesignerAudioProcessorEditor::setEqMode()
+void PolarDesignerAudioProcessorEditor::updateUIForEQMode()
 {
     int activeIdx = polarDesignerProcessor.getEqState();
     if (activeIdx == 0)
diff --git a/Source/PluginEditor.h b/Source/PluginEditor.h
index 61f9328..589c96c 100644
--- a/Source/PluginEditor.h
+++ b/Source/PluginEditor.h
@@ -68,7 +68,6 @@ public:
     void buttonClicked (Button* button) override;
     void sliderValueChanged (Slider* slider) override;
 
-    void setEqMode();
     void calculateLockedBands(int nBands, bool trimSliderIncr);
     float getABButtonAlphaFromLayerState(int layerState);
     // Helper method to calculate flex on the base of bandlimitPathComponents
@@ -193,14 +192,12 @@ private:
 #endif
 
     //==========================================================================
-    void nEditorBandsChanged();
     void loadFile();
     void saveFile();
     void timerCallback() override;
     bool getSoloActive();
-    void setMainAreaEnabled(bool enable);
-    void enableLatencyUI (bool set);
-    void zeroLatencyModeChanged();
+
+
     void showPresetList(bool shouldShow);
     void setBandEnabled(int bandNr, bool enable);
 
@@ -209,6 +206,22 @@ private:
 
     void mouseDown(const MouseEvent& event) override;
 
+
+    /* Sets the elements of the MainUI Active as necessary */
+    void setMainUIEnabled (bool enabled);
+
+    /* Sets the elements of the UI relevant to Zero Latency Mode as necessary */
+    void setLatencyUIEnabled (bool enabled);
+
+    /* Update the UI to reflect the state of Zero Latency Mode */
+    void updateUIForZeroLatencyMode();
+
+    /* Update the UI to reflect the number of active EQ Bands */
+    void updateUIForNumberActiveEQBands();
+
+    /* Update the UI to reflect the current EQ Mode */
+    void updateUIForEQMode();
+
 #ifdef AA_INCLUDE_MELATONIN
     melatonin::Inspector inspector { *this, false };
 #endif
diff --git a/Source/PluginProcessor.cpp b/Source/PluginProcessor.cpp
index d0400fd..abba668 100644
--- a/Source/PluginProcessor.cpp
+++ b/Source/PluginProcessor.cpp
@@ -88,10 +88,6 @@ PolarDesignerAudioProcessor::PolarDesignerAudioProcessor() : AudioProcessor(Buse
                                                             disturberRecorded(false),
                                                             signalRecorded(false),
                                                             currentSampleRate(48000) {
-#if PERFETTO
-    MelatoninPerfetto::get().beginSession();
-#endif
-
     vtsParams.addParameterListener("trimPosition", this);
 
     vtsParams.addParameterListener("xOverF1", this);
@@ -160,9 +156,6 @@ PolarDesignerAudioProcessor::PolarDesignerAudioProcessor() : AudioProcessor(Buse
 }
 
 PolarDesignerAudioProcessor::~PolarDesignerAudioProcessor() {
-#if PERFETTO
-    MelatoninPerfetto::get().endSession();
-#endif
 }
 
 //==============================================================================
@@ -566,9 +559,6 @@ void PolarDesignerAudioProcessor::setStateInformation(const void *data, int size
  * * set atomic flags for processBlock to check at the start of each block
 */
 void PolarDesignerAudioProcessor::parameterChanged(const String &parameterID, float newValue) {
-
-    TRACE_EVENT("dsp", "parameterChanged", "paramID:", perfetto::DynamicString{parameterID.toStdString()});
-
     if (parameterID.startsWith("trimSlider")) {
         // trimPosition->load() = newValue; // TODO: update trimSlider position according to automation
     } else if (parameterID == "nrBands") {
@@ -593,8 +583,6 @@ void PolarDesignerAudioProcessor::parameterChanged(const String &parameterID, fl
         setProxCompCoefficients(proxDistance->load());
     } else if (parameterID == "zeroLatencyMode") {
 
-        TRACE_EVENT("dsp", "zeroLatencyMode");
-
         updateLatency();
 
         // zero Latency Mode turned off
@@ -625,17 +613,16 @@ void PolarDesignerAudioProcessor::parameterChanged(const String &parameterID, fl
                 oldNrBandsA = nProcessorBandsPtr->load();
             }
 
-            // If zeroLatencyMode, nrBands should be 1
-            vtsParams.getParameter("nrBands")->setValueNotifyingHost(
-                            vtsParams.getParameter("nrBands")->convertTo0to1(1));
-
             vtsParams.getParameter("proximity")->setValueNotifyingHost(
                     vtsParams.getParameter("proximity")->convertTo0to1(0));
 
             zeroLatencyModeChanged = true;
         }
 
-
+        if (zeroLatencyModeChanged.get()) {
+            vtsParams.state.setProperty("oldZeroLatencyMode", var(zeroLatencyMode->load()), nullptr);
+            vtsParams.state.setProperty("zeroLatencyMode", var(zeroLatencyMode->load()), nullptr);
+        }
 
     } else if (parameterID == "syncChannel" && syncChannelPtr->load() >= 0.5f) {
         int ch = (int) syncChannelPtr->load() - 1;
@@ -1401,9 +1388,6 @@ void PolarDesignerAudioProcessor::setProxCompCoefficients(float distance) {
 }
 
 void PolarDesignerAudioProcessor::timerCallback() {
-
-    TRACE_DSP();
-
     if (syncChannelPtr->load() > 0.5f) {
         readingSharedParams = true;
 
@@ -1448,6 +1432,14 @@ void PolarDesignerAudioProcessor::timerCallback() {
                     vtsParams.getParameterRange("proximity").convertTo0to1(paramsToSync.proximity));
         }
 
+        // !J! Disabled while debugging sync issues
+        float _zlm = zeroLatencyMode->load();
+
+        if (paramsToSync.zeroLatencyMode != (_zlm > 0.f)) {
+            vtsParams.getParameter ("zeroLatencyMode")->setValueNotifyingHost (vtsParams.getParameterRange ("zeroLatencyMode").convertTo0to1 (paramsToSync.zeroLatencyMode));
+        }
+
+
         if (!exactlyEqual(allowBackwardsPattern->load(), paramsToSync.allowBackwardsPattern ? 1.0f : 0.0f))
         {
             vtsParams.getParameter ("allowBackwardsPattern")->setValueNotifyingHost (vtsParams.getParameterRange ("allowBackwardsPattern").convertTo0to1 (paramsToSync.allowBackwardsPattern));
@@ -1463,11 +1455,6 @@ void PolarDesignerAudioProcessor::timerCallback() {
             ffDfEqChanged = true;
         }
 
-        if (zeroLatencyMode->load() != paramsToSync.zeroLatencyMode) {
-            vtsParams.getParameter("zeroLatencyMode")->setValueNotifyingHost(
-                vtsParams.getParameterRange("zeroLatencyMode").convertTo0to1(paramsToSync.zeroLatencyMode));
-        }
-
         readingSharedParams = false;
     }
 }
diff --git a/Source/PluginProcessor.h b/Source/PluginProcessor.h
index ca293ce..a3d66c2 100644
--- a/Source/PluginProcessor.h
+++ b/Source/PluginProcessor.h
@@ -32,7 +32,6 @@
 
 #include <cfloat>
 
-#include "melatonin_perfetto/melatonin_perfetto.h"
 
 /* PolarDesigner has a maximum of 5 EQ's .. */
 static const int MAX_NUM_EQS = 5;
@@ -207,11 +206,6 @@ public:
     float hzToZeroToOne(int idx, float hz);
     float hzFromZeroToOne(int idx, float val);
 
-#if PERFETTO
-    // perfetto
-    std::unique_ptr<perfetto::TracingSession> tracingSession;
-#endif
-
     bool zeroLatencyModeActive() {
         bool _zd = (zeroLatencyMode->load() > 0.5f);
         return (_zd);
diff --git a/modules/melatonin_perfetto b/modules/melatonin_perfetto
deleted file mode 160000
index 1fdceed..0000000
--- a/modules/melatonin_perfetto
+++ /dev/null
@@ -1 +0,0 @@
-Subproject commit 1fdceed03b3f83c983ab759e9b3e642792c0d0ed
diff --git a/resources/customComponents/MultiTextButton.h b/resources/customComponents/MultiTextButton.h
index fc73531..b23dd34 100644
--- a/resources/customComponents/MultiTextButton.h
+++ b/resources/customComponents/MultiTextButton.h
@@ -62,6 +62,17 @@ public:
         resized();
     }
 
+    int getButtonsNumber()
+    {
+        for (int i = 0; i < textButtonArray.size(); i++)
+        {
+            if (textButtonArray.getUnchecked(i)->getToggleState())
+                return (i);
+        }
+
+        return 0;
+    }
+
     void disableAllButtons() {
         for (int i = 0; i < textButtonArray.size(); i++)
         {
